service discovery proposal using kubernetes service annotations

KurtStam opened the pull request:
Note that the annotation's namespace of 'apiman.io' could be changed to 'kubernetes.io'.  See also
https://github.com/kubernetes/kubernetes/issues/14828

Commit:
service discovery proposal using kubernetes service annotations

KurtStam commented:
I signed it!

bgrant0607 commented:
cc @thockin @smarterclayton

bgrant0607 commented:
Ref #17543

thockin commented on docs/proposals/service-discovery.md:
```diff
@@ -0,0 +1,102 @@
+
+## Goal of this document
+
+To consume a service, a developer needs to know the full URL and a description of the API. Kubernetes contains the host and port information of a service, but it lacks the scheme and the pathe information needed if the services is not bound at the root. In this document we propose some standard kubernetes service annotations to fix these gaps. It is important that these annotations are a standard ...
```
s/pathe/path

thockin replied:
s/services/service

thockin commented on docs/proposals/service-discovery.md:
```diff
@@ -0,0 +1,102 @@
+
+'apiman.io/servicepath' - the path part of the service endpoint url. An example value could be 'cxfcdi',
+
+'apiman.io/servicescheme' - the scheme part of the service endpoint url. An example values could be 'http' or 'https'.
```
s/An examples/Some examples

thockin replied:
service-scheme

thockin commented on docs/proposals/service-discovery.md:
```diff
@@ -0,0 +1,102 @@
+
+Kubernetes allows the creation of Service Annotations. Here we propose the use of the following standard annotations
+
+'apiman.io/servicepath' - the path part of the service endpoint url. An example value could be 'cxfcdi',
```
service-path ?

KurtStam replied:
Sure - I like adding the hyphen.

thockin commented on docs/proposals/service-discovery.md:
```diff
@@ -0,0 +1,102 @@
+
+'apiman.io/servicescheme' - the scheme part of the service endpoint url. An example values could be 'http' or 'https'.
+
+'apiman.io/servicetype' - the protocol of the service. Example values can be 'SOAP' or 'REST',
```
service-type, but can we please find a better word that "type"?  We already literally have a "type" field.

KurtStam replied:
How about 'service-style' or 'service-arch-style' as 'protocol' seems too strict and already used as well. I could see 'service-api-protocol' though, as the 'protocol' references to the _wire_-protocol.

thockin replied:
I can live with either proposal, but "arch" is not a great abbreviation, so I prefer api-protocol

thockin commented on docs/proposals/service-discovery.md:
```diff
@@ -0,0 +1,102 @@
+
+'apiman.io/servicetype' - the protocol of the service. Example values can be 'SOAP' or 'REST',
+
+'apiman.io/descriptionpath' - the path part of the service description document’s endpoint. It is a pretty safe assumption that the service self documents. An example value for a swagger 2.0 document can be 'cxfcdi/swagger.json',
```
description-path

thockin commented on docs/proposals/service-discovery.md:
```diff
@@ -0,0 +1,102 @@
+
+'apiman.io/descriptionpath' - the path part of the service description document’s endpoint. It is a pretty safe assumption that the service self documents. An example value for a swagger 2.0 document can be 'cxfcdi/swagger.json',
+
+'apiman.io/descriptiontype' - the type of Description Language used. Example values can be 'WSDL', 'WADL', 'SwaggerJSON', 'SwaggerYAML'.
```
descritption-language or description-format or description-syntax

KurtStam replied:
I like description-language, to stay inline with the '*DL' part of WSDL and WADL.

thockin commented:
If we're going to call them standard, we should just use the `kubernetes.io` namespace for the annotations.  If these are primarily consumed by a single other system, then a namespace for that system makes sense.  But it sounds like you're proposing these as general-purpose.  I can live with that.

KurtStam commented:
@thockin Yes I would prefer using _kubernetes.io_ to emphasize that it not specific to apiman.io, but generally usable. Acceptance of this standard would be beneficial for everyone using Kubernetes.

thockin commented:
I had a quick conversation and Brian pointed out that we have convention in
docs/devel/api-conventions.md that applies here.  Structure goes to the left.
These are all Service-specific so:

```
service.kubernetes.io/path
service.kubernetes.io/scheme
service.kubernetes.io/api-protocol
service.kubernetes.io/description-path
service.kubernetes.io/description-language
```

Now we can argue if those are clear enough.

Should `path` be `http-path`?  The scheme suggests non-HTTP schemes - does path
apply to those (e.g. jdbc)?

Are description-path and -language better named as api-description- ?

Should these all be `api.service.kubernetes.io/` prefixed?

Are these going to apply Ingress as well as Service?  if so, maybe that affects
final name.

KurtStam commented:
Answering your questions in reverse order. I don't think these annotations apply to Ingress. Correct me if I'm wrong but at the moment Ingress is defined sort as 'mod_proxy' to consolidate endpoints, and forward traffic to the actual services based on a path based rule. Now, the service developer knows where his service api and the DL doc lives, and uses the proposed annotations to capture this info. The ingress simply _maps_ it. (It begs the question if the Ingres API itself should have easy methods that map the api location and api description document, but that is unrelated to this proposal)

Next, now that we appearantly are talking about the API on a service, maybe the annotation names should be

```
api.kubernetes.io/path
api.kubernetes.io/scheme
api.kubernetes.io/protocol
description.api.kubernetes.io/path
description.api.kubernetes.io/language
```

Finally, I think the scheme will be used for other types of URLs besides http, though that will be probably be 90% of all use cases. It's useful even to specify https vs http.

thockin commented:
> Next, now that we appearantly are talking about the API on a service, maybe the annotation names should be
>
> api.kubernetes.io/path
> api.kubernetes.io/scheme
> api.kubernetes.io/protocol
> description.api.kubernetes.io/path
> description.api.kubernetes.io/language

I think "api" is too generic.  `api.service.kubernetes.io/path` or
`service.kubernetes.io/api-path`.  I have a slight preference for the
latter - the namespace is "about services", and the datum is "the api
path"

> Finally, I think the scheme will be used for other types of URLs besides http, though that will be probably be 90% of all use cases. It's useful even to specify https vs http.

Will `path` apply to non-HTTP `scheme`s?

KurtStam commented:
> I think "api" is too generic.  `api.service.kubernetes.io/path` or
> `service.kubernetes.io/api-path`.  I have a slight preference for the
> latter - the namespace is "about services", and the datum is "the api
> path"

You have a point there. I just like how 'api.kubernetes.io' brings down the right-hand sides to one word ('protocol' in particular), I like the symmetry and how it mashes with 'description.api.kubernetes.io' But, like I said, you do have point, so if you feel strongly that 'service.kubernetes.io' makes more sense that I'm fine with that.

> Will`path` apply to non-HTTP `scheme`s?

Yes the generic URL definition is:
`scheme:[//[user:password@]host[:port]][/]path[?query][#fragment]`
and some examples of non http use are:
- DB:   'jdbc:postgresql://host:port/database'
- FTP: 'ftp://kurt:secret@www.myserver.com/project/2016/'
- IRC: 'irc://irc.example.com/channel'

I was going for 'keeping things simple' but should we support 'user' and 'password'?; Sometimes completeness makes things simpler to understand and use. I don't see the need for 'query' or 'fragment' for this ever really.

thockin commented:
embedding user and password here is not a good idea, I think.  This is what
the service linking stuff is supposed to do.  Let's at least wait until
that settles.

Ping me when all ready for re-review.

On Fri, Jan 29, 2016 at 7:14 AM, Kurt Stam <notifications@github.com> wrote:

> I think "api" is too generic. api.service.kubernetes.io/path or
> service.kubernetes.io/api-path. I have a slight preference for the
> latter - the namespace is "about services", and the datum is "the api
> path"
>
> You have a point there. I just like how 'api.kubernetes.io' brings down
> the right-hand sides to one word ('protocol' in particular), I like the
> symmetry and how it mashes with 'description.api.kubernetes.io' But, like
> I said, you do have point, so if you feel strongly that '
> service.kubernetes.io' makes more sense that I'm fine with that.
>
> Willpath apply to non-HTTP schemes?
>
> Yes the generic URL definition is:
> scheme:[//[user:password@]host[:port]][/]path[?query][#fragment]
> and some examples of non http use are:
>
>    - DB: 'jdbc:postgresql://host:port/database'
>    - FTP: 'ftp://kurt:secret@www.myserver.com/project/2016/'
>    - IRC: 'irc://irc.example.com/channel'
>
> I was going for 'keeping things simple' but should we support 'user' and
> 'password'?; Sometimes completeness makes things simpler to understand and
> use. I don't see the need for 'query' or 'fragment' for this ever really.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20216#issuecomment-176803033>
> .
>

KurtStam commented:
@thockin, Tim I updated the proposal with the comments in the review and I fixed up some formatting. I hope I convinced you on 'api.kubernetes.io'. if not, just let me know and I'll fix it up. Happy weekend! --Kurt

thockin commented on docs/proposals/service-discovery.md:
```diff
@@ -0,0 +1,98 @@
+      "apiVersion" : "v1",
+      "kind" : "Service",
+      "metadata" : {
+        "annotations" : {
```
Didn't convince, sorry.  `api` makes it appear to be about the kube api.

How about, final answer:
```
         "api.service.kubernetes.io/protocol" : "REST",
          "api.service.kubernetes.io/scheme" "http",
          "api.service.kubernetes.io/path" : "cxfcdi",
          "api.service.kubernetes.io/description-path" : "cxfcdi/swagger.json",
          "api.service.kubernetes.io/description-language" : "SwaggerJSON"
```

thockin commented on docs/proposals/service-discovery.md:
```diff
@@ -0,0 +1,98 @@
+
+* `api.kubernetes.io/path` - the path part of the service endpoint url. An example value could be `cxfcdi`,
+* `api.kubernetes.io/scheme` - the scheme part of the service endpoint url. Some values could be `http` or `https`.
+* `api.kubernetes.io/prototype` - the protocol of the service. Example values can be `SOAP` or `REST`,
```
Do you want to document SOAP and REST as known values, so that anyone who uses this will choose them instead of Soap or soap or rest or Rest?

thockin commented on docs/proposals/service-discovery.md:
```diff
@@ -0,0 +1,98 @@
+* `api.kubernetes.io/scheme` - the scheme part of the service endpoint url. Some values could be `http` or `https`.
+* `api.kubernetes.io/prototype` - the protocol of the service. Example values can be `SOAP` or `REST`,
+* `description.api.kubernetes.io/path` - the path part of the service description document’s endpoint. It is a pretty safe assumption that the service self-documents. An example value for a swagger 2.0 document can be `cxfcdi/swagger.json`,
+* `description.api.kubernetes.io/language` - the type of Description Language used. Example values can be `WSDL`, `WADL`, `SwaggerJSON`, `SwaggerYAML`.
```
As above, do you want to document these as known values so nobody uses "Swagger" or "Swagger-JSON" when they should have used "SwaggerJSON"?

KurtStam commented:
@thockin - Great suggestions, I incorporated them in commit "final answer" - cacc707: https://github.com/KurtStam/kubernetes/commit/cacc70741219aa3f2e1ed29c44c430472b8a4108

thockin commented:
LGTM.  Please squash

KurtStam commented:
squashed
